File: java-solver/src/fi/spektrum/sudoku/dlx/DLXParser.java
package fi.spektrum.sudoku.dlx;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 *
 * @author Lasse Lybeck
 */
public class DLXParser {

    private final int[][] sudoku;
    private final int n;
    private final int n2;
    private final int n4;
    private final int cols;
    private final ColumnNode header;
    private Map<Integer, ColumnNode> columns;

    public DLXParser(int[][] sudoku) {
        this.sudoku = sudoku;
        n2 = sudoku.length;
        n = (int) Math.sqrt(n2);
        n4 = n2 * n2;
        cols = 4 * n2 * n2;
        header = parseSudoku();
    }

    public ColumnNode getHeader() {
        return header;
    }

    public final ColumnNode parseSudoku() {

        ColumnNode h = initHeader();
        columns = initMap(h);

        for (int i = 0; i < sudoku.length; ++i) {
            for (int j = 0; j < sudoku[i].length; ++j) {
                int num = sudoku[i][j];
                if (num != 0) {
                    addRow(i, j, num);
                } else {
                    for (int k = 1; k <= n2; k++) {
                        addRow(i, j, k);
                    }
                }
            }
        }

        return h;
    }

    private ColumnNode initHeader() {
        ColumnNode h = new ColumnNode();
        h.setRight(h);
        h.setLeft(h);
        h.setUp(h);
        h.setDown(h);
        h.setColumn(h);
        return h;
    }

    private int[][] getDLXMatrix(int[][] sudoku) {
        List<int[]> list = new ArrayList<>();
        for (int i = 0; i < sudoku.length; ++i) {
            for (int j = 0; j < sudoku[i].length; ++j) {
                int num = sudoku[i][j];
                if (num != 0) {
                    list.add(makeRow(i, j, num));
                } else {
                    for (int k = 1; k <= n2; k++) {
                        list.add(makeRow(i, j, k));
                    }
                }
            }
        }
        int[][] mat = list.toArray(new int[list.size()][cols]);
        return mat;
    }

    private int[] makeRow(int i, int j, int num) {
        int[] row = new int[cols];
        int ind = n2 * i + j;
        row[ind] = num;
        row[n4 + n2 * i + (num - 1)] = num;
        row[2 * n4 + n2 * j + (num - 1)] = num;
        row[3 * n4 + n2 * getBoxNumber(i, j, n) + (num - 1)] = num;
        return row;
    }

    private int getBoxNumber(int i, int j, int n) {
        int ib = i / n;
        int jb = j / n;
        return ib * n + jb;
    }

    private void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            for (int i : row) {
                if (i != 0) {
                    System.out.print(i);
                } else {
                    System.out.print(" ");
                }
            }
            System.out.println();
        }
    }

    private Map<Integer, ColumnNode> initMap(ColumnNode h) {
        Map<Integer, ColumnNode> map = new HashMap<>(cols);
        ColumnNode prev = h;
        ColumnNode curr;
        for (int i = 0; i < cols; i++) {
            curr = new ColumnNode();
            curr.setUp(curr);
            curr.setDown(curr);
            curr.setLeft(prev);
            prev.setRight(curr);
            map.put(i, curr);
            prev = curr;
        }
        prev.setRight(h);
        h.setLeft(prev);
        return map;
    }

    private void addRow(int i, int j, int num) {
        Node[] nodes = getNodes(num);
        int[] indices = getIndices(i, j, num);
        addRowToSparseMatrix(nodes, indices);
    }

    private Node[] getNodes(int num) {
        Node[] nodes = initNodes(num);
        Node prev = nodes[0];
        Node curr;
        for (int i = 1; i < nodes.length; i++) {
            curr = nodes[i];
            curr.setLeft(prev);
            prev.setRight(curr);
            prev = curr;
        }
        prev.setRight(nodes[0]);
        nodes[0].setLeft(prev);
        return nodes;
    }

    private Node[] initNodes(int num) {
        Node[] nodes = new Node[4];
        nodes[0] = new Node();
        nodes[1] = new Node();
        nodes[2] = new Node();
        nodes[3] = new Node();
        for (Node node : nodes) {
            node.setValue(num);
        }
        return nodes;
    }

    protected int[] getIndices(int i, int j, int num) {
        int[] inds = new int[4];
        inds[0] = n2 * i + j;
        inds[1] = n4 + n2 * i + (num - 1);
        inds[2] = 2 * n4 + n2 * j + (num - 1);
        inds[3] = 3 * n4 + n2 * getBoxNumber(i, j, n) + (num - 1);
        return inds;
    }

    private void addRowToSparseMatrix(Node[] nodes, int[] indices) {
        for (int i = 0; i < nodes.length; i++) {
            addNodeToSparseMatrix(nodes[i], indices[i]);
        }
    }

    private void addNodeToSparseMatrix(Node node, int index) {
        ColumnNode column = columns.get(index);
        column.getUp().setDown(node);
        node.setUp(column.getUp());
        column.setUp(node);
        node.setDown(column);
        column.setSize(column.getSize() + 1);
    }
}

File: cpp-solver/dlx-parser.cpp

#include <iostream>

#include "dlx-parser.h"

namespace dlx {

    dlx_parser::dlx_parser(sudoku s, int n) : sudoku_(s), n_(n) {
        n2_ = n_ * n_;
        n4_ = n2_ * n2_;
        cols_ = 4 * n4_;
        columns_ = new column_node*[cols_];
        nodes_ = new node*[4];
        indices_ = new int[4];
    }

    dlx_parser::~dlx_parser() {
        delete columns_;
        delete indices_;
        delete nodes_;
    }

    column_node* dlx_parser::parse() {

#ifndef NDEBUG
        std::cout << "Parsing..." << std::endl;
#endif

        column_node* h = init_header();

        init_columns(h);

        add_rows();

        return h;
    }

    column_node* dlx_parser::init_header() {

#ifndef NDEBUG
        std::cout << "  init_header" << std::endl;
#endif
        column_node* h = new column_node();
        L(h) = h;
        R(h) = h;
        return h;
    }

    void dlx_parser::init_columns(column_node* h) {

#ifndef NDEBUG
        std::cout << "  init_columns" << std::endl;
#endif
        column_node* prev = h;
        column_node* curr;
        for (int i = 0; i < cols_; i++) {
            curr = new column_node();
            U(curr) = curr;
            D(curr) = curr;
            L(curr) = prev;
            R(prev) = curr;
            columns_[i] = curr;
            prev = curr;
        }
        R(prev) = h;
        L(h) = prev;
    }

    void dlx_parser::add_rows() {

#ifndef NDEBUG
        std::cout << "  add_rows" << std::endl;
#endif
        int num;
        for (int i = 0; i < n2_; i++) {
            for (int j = 0; j < n2_; j++) {
                num = get(sudoku_, n2_, i, j);
                if (num != 0) {
                    add_row(i, j, num);
                } else {
                    for (int k = 0; k < n2_; k++) {
                        add_row(i, j, k);
                    }
                }
            }
        }
    }

    void dlx_parser::add_row(int i, int j, int num) {
        setup_nodes(i, j, num);
        setup_indices(i, j, num);
        add_row_to_matrix();
    }

    void dlx_parser::setup_nodes(int i, int j, int num) {

        // create the nodes
        int index = get_index(n2_, i, j);
        for (int i = 0; i < 4; i++) {
            nodes_[i] = new node();
            V(nodes_[i]) = num;
            I(nodes_[i]) = index;
        }

        // link the nodes together
        node* prev = nodes_[0];
        node* curr;
        for (int k = 1; k < 4; k++) {
            curr = nodes_[k];
            L(curr) = prev;
            R(prev) = curr;
            prev = curr;
        }
        R(prev) = nodes_[0];
        L(nodes_[0]) = prev;
    }

    void dlx_parser::setup_indices(int i, int j, int num) {
        indices_[0] = 0 * n4_ + n2_ * i + j;
        indices_[1] = 1 * n4_ + n2_ * i + (num - 1);
        indices_[2] = 2 * n4_ + n2_ * j + (num - 1);
        indices_[3] = 3 * n4_ + n2_ * get_box_number(i, j) + (num - 1);
    }

    int dlx_parser::get_box_number(int i, int j) {
        int ib = i / n_;
        int jb = j / n_;
        return ib * n_ + jb;
    }

    void dlx_parser::add_row_to_matrix() {
        node* n;
        int index;
        column_node* c;
        for (int i = 0; i < 4; i++) {
            n = nodes_[i];
            index = indices_[i];
            c = columns_[index];
            D(U(c)) = n;
            U(n) = U(c);
            U(c) = n;
            D(n) = c;
            C(n) = c;
            ++S(c);
        }
    }
}
